C++20 feature: latch

Issue tracking implementation of `latch`es, introduced in C++20.

See [latch](https://en.cppreference.com/w/cpp/thread/latch) on cppreference.com.

Potential implementations:
- Atomic-based. Portable, but does not cause threads to sleep when blocking.
- Atomic variable & `condition_variable`. Mostly efficient. May have issues with cost of construction / destruction.
- Atomic & [WaitOnAddress](https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitonaddress). Requires Windows 8, but should have minimal overhead.

https://www.modernescpp.com/index.php/latches-in-c-20
I have  question. When I use C++2a commands in gcc 11.3.0 I have info that gcc isn't `std::latch`.
```
workers.cpp:8:6: error: 'latch' in namespace 'std' does not name a type
    8 | std::latch workDone(6);
      |      ^~~~~
```
When I use 'latch.h' gcc 11.3.0 and above hasn't problem.
https://github.com/luncliff/latch

--- a/cmake_stdheaders_generator/CMakeLists.txt
+++ b/cmake_stdheaders_generator/CMakeLists.txt
@@ -61,6 +61,8 @@ endif()
 generate_mingw_stdthreads_header(condition_variable "${MINGW_STDTHREADS_DIR}")
 # <future>
 generate_mingw_stdthreads_header(future "${MINGW_STDTHREADS_DIR}")
+# <latch>
+generate_mingw_stdthreads_header(latch "${MINGW_STDTHREADS_DIR}")
 # <mutex>
 generate_mingw_stdthreads_header(mutex "${MINGW_STDTHREADS_DIR}")
 # <shared_mutex>

--- a/mingw.latch.h
+++ b/mingw.latch.h
@@ -27,11 +27,6 @@
 #error The contents of <latch> require a C++20 compiler!
 #endif
 
-#include <sdkddkver.h>    //  Detect Windows version
-#if !defined(_WIN32_WINNT) || (_WIN32_WINNT < 0x0501)
-#error To use the MinGW-std-threads library, you will need to define the macro _WIN32_WINNT to be 0x0501 (Windows XP) or higher.
-#endif
-
 #include <atomic>
 #include <cassert>        // for descriptive errors
 #include <cstddef>        // for std::ptrdiff_t
@@ -47,8 +42,9 @@ public:
         return std::numeric_limits<std::ptrdiff_t>::max();
     }
 
-    constexpr explicit latch(std::ptrdiff_t expected) : mCounter{expected}
+    constexpr explicit latch(std::ptrdiff_t expected) noexcept : mCounter{expected}
     {
+        assert(expected >= 0);
     }
 
     ~latch()=default;
@@ -59,8 +55,11 @@ public:
     {
         assert(update >= 0);
 
-        const auto current = mCounter.fetch_sub(update, std::memory_order_release) - update;
-        if (current == 0)
+        const auto current = mCounter.fetch_sub(update, std::memory_order_release);
+
+        assert(update <= current);
+
+        if (current <= update)
         {
             mCounter.notify_all();
         }
@@ -68,26 +67,36 @@ public:
 
     [[nodiscard]] bool try_wait() const noexcept
     {
-        return 0 == mCounter.load(std::memory_order_acquire);
+        return mCounter.load(std::memory_order_acquire) <= 0;
     }
 
+    /**
+    * The call to atomic::wait() may unblock with a non-zero counter in some edge cases (see GH-91 for an in-depth explanation)
+    * To address the edge cases this loop will continue to wait until the counter has been verified to have reached a non-positive value
+    */         
     void wait() const
     {
-        const auto current = mCounter.load(std::memory_order_acquire);
-        if (current == 0)
+        while (true)
         {
-            return;
-        }
+            const auto current = mCounter.load(std::memory_order_acquire);
+            if (current < 0)
+            {
+                return;
+            }
 
-        mCounter.wait(current, std::memory_order_relaxed);
+            mCounter.wait(current, std::memory_order_relaxed);
+        }
     }
 
     void arrive_and_wait(const std::ptrdiff_t update = 1) noexcept
     {
         assert(update >= 0);
 
-        const auto current = mCounter.fetch_sub(update, std::memory_order_acq_rel) - update;
-        if (current == 0)
+        const auto current = mCounter.fetch_sub(update, std::memory_order_acq_rel);
+
+        assert(current <= update);
+
+        if (current == update)
         {
             mCounter.notify_all();
         }
@@ -95,7 +104,7 @@ public:
         {
             assert(current > 0);
 
-            mCounter.wait(current, std::memory_order_relaxed);
+            mCounter.wait(current - update, std::memory_order_relaxed);
             wait();
         }
     }

--- a/mingw.latch.h
+++ b/mingw.latch.h
@@ -72,14 +72,14 @@ public:
 
     /**
     * The call to atomic::wait() may unblock with a non-zero counter in some edge cases (see GH-91 for an in-depth explanation)
-    * To address the edge cases this loop will continue to wait until the counter has been verified to have reached a non-positive value
+    * To address the edge cases this loop will continue to wait until the counter has been verified to have reached 0 (or less)
     */         
     void wait() const
     {
         while (true)
         {
             const auto current = mCounter.load(std::memory_order_acquire);
-            if (current < 0)
+            if (current <= 0)
             {
                 return;
             }

--- a/mingw.latch.h
+++ b/mingw.latch.h
@@ -1,0 +1,113 @@
+/// \file mingw.latch.h
+/// \brief std::latch implementation for MinGW.
+///
+/// (c) 2023 by Mega Limited, Auckland, New Zealand
+/// \author Edoardo Sanguineti
+///
+/// \copyright Simplified (2-clause) BSD License.
+///
+/// \note This file may become part of the mingw-w64 runtime package. If/when
+/// this happens, the appropriate license will be added, i.e. this code will
+/// become dual-licensed, and the current BSD 2-clause license will stay.
+
+//  Notes on the namespaces:
+//  - The implementation can be accessed directly in the namespace
+//    mingw_stdthread.
+//  - Objects will be brought into namespace std by a using directive. This
+//    will cause objects declared in std (such as MinGW's implementation) to
+//    hide this implementation's definitions.
+//  The end result is that if MinGW supplies an object, it is automatically
+//  used. If MinGW does not supply an object, this implementation's version will
+//  instead be used.
+
+#ifndef MINGW_LATCH_H_
+#define MINGW_LATCH_H_
+
+#if !defined(__cplusplus) || (__cplusplus < 202002L)
+#error The contents of <latch> require a C++20 compiler!
+#endif
+
+#include <sdkddkver.h>    //  Detect Windows version
+#if !defined(_WIN32_WINNT) || (_WIN32_WINNT < 0x0501)
+#error To use the MinGW-std-threads library, you will need to define the macro _WIN32_WINNT to be 0x0501 (Windows XP) or higher.
+#endif
+
+#include <atomic>
+#include <cassert>        // for descriptive errors
+#include <cstddef>        // for std::ptrdiff_t
+#include <limits>         // for std::numeric_limits
+
+namespace mingw_stdthread
+{
+class latch
+{
+public:
+    [[nodiscard]] static constexpr std::ptrdiff_t max() noexcept
+    {
+        return std::numeric_limits<std::ptrdiff_t>::max();
+    }
+
+    constexpr explicit latch(std::ptrdiff_t expected) : mCounter{expected}
+    {
+    }
+
+    ~latch()=default;
+    latch(const latch&)=delete;
+    latch& operator=(const latch&)=delete;
+
+    void count_down(std::ptrdiff_t update = 1 )
+    {
+        assert(update >= 0);
+
+        const auto current = mCounter.fetch_sub(update, std::memory_order_release) - update;
+        if (current == 0)
+        {
+            mCounter.notify_all();
+        }
+    }
+
+    [[nodiscard]] bool try_wait() const noexcept
+    {
+        return 0 == mCounter.load(std::memory_order_acquire);
+    }
+
+    void wait() const
+    {
+        const auto current = mCounter.load(std::memory_order_acquire);
+        if (current == 0)
+        {
+            return;
+        }
+
+        mCounter.wait(current, std::memory_order_relaxed);
+    }
+
+    void arrive_and_wait(const std::ptrdiff_t update = 1) noexcept
+    {
+        assert(update >= 0);
+
+        const auto current = mCounter.fetch_sub(update, std::memory_order_acq_rel) - update;
+        if (current == 0)
+        {
+            mCounter.notify_all();
+        }
+        else
+        {
+            assert(current > 0);
+
+            mCounter.wait(current, std::memory_order_relaxed);
+            wait();
+        }
+    }
+
+private:
+    std::atomic<std::ptrdiff_t> mCounter;
+};
+} //  Namespace mingw_stdthread
+
+namespace std
+{
+using mingw_stdthread::latch;
+} //  Namespace std
+
+#endif // MINGW_LATCH_H_

--- a/mingw.latch.h
+++ b/mingw.latch.h
@@ -51,7 +51,7 @@ public:
     latch(const latch&)=delete;
     latch& operator=(const latch&)=delete;
 
-    void count_down(std::ptrdiff_t update = 1 )
+    void count_down(std::ptrdiff_t update = 1)
     {
         assert(update >= 0);
 

--- a/tests/tests.cpp
+++ b/tests/tests.cpp
@@ -4,12 +4,14 @@
   #include <mingw.condition_variable.h>
   #include <mingw.shared_mutex.h>
   #include <mingw.future.h>
+  #include <mingw.latch.h>
 #else
   #include <thread>
   #include <mutex>
   #include <condition_variable>
   #include <shared_mutex>
   #include <future>
+  #include <latch>
 #endif
 #include <atomic>
 #include <cassert>
@@ -293,6 +295,38 @@ void test_future ()
   test_future_get_value(async_member);
 }
 
+void test_latch ()
+{
+  std::latch latch(5);
+
+  std::thread thread1([&latch] { latch.wait(); });
+  std::thread thread2([&latch] { latch.arrive_and_wait(2); });
+
+  latch.count_down();
+
+  log("Testing count_down...");
+  latch.count_down(2);
+
+  thread1.join();
+  thread2.join();
+}
+
+void test_latch_release_wait ()
+{
+  std::latch latch(5);
+
+  std::thread thread1([&latch] { latch.wait(); });
+  std::thread thread2([&latch] { latch.arrive_and_wait(2); });
+
+  latch.count_down();
+
+  log("Testing release_wait...");
+  latch.arrive_and_wait(2);
+
+  thread1.join();
+  thread2.join();
+}
+
 #define TEST_SL_MV_CPY(ClassName) \
     static_assert(std::is_standard_layout<ClassName>::value, \
                   "ClassName does not satisfy concept StandardLayoutType."); \
@@ -480,6 +514,12 @@ int main()
       log("Testing <future>'s use of allocators. Should allocate, then deallocate.");
       promise<int> allocated_promise (std::allocator_arg, CustomAllocator<unsigned>());
       allocated_promise.set_value(7);
+    }
+
+    {
+      log("Testing implementation of <latch>...");
+      test_latch();
+      test_latch_release_wait();
     }
 
     return failure_count.load() != 0;

--- a/tests/tests.cpp
+++ b/tests/tests.cpp
@@ -4,14 +4,22 @@
   #include <mingw.condition_variable.h>
   #include <mingw.shared_mutex.h>
   #include <mingw.future.h>
-  #include <mingw.latch.h>
+
+  #if defined(__cplusplus) && (__cplusplus >= 202002L)
+    #include <mingw.latch.h>
+  #endif
+
 #else
   #include <thread>
   #include <mutex>
   #include <condition_variable>
   #include <shared_mutex>
   #include <future>
-  #include <latch>
+
+  #if defined(__cplusplus) && (__cplusplus >= 202002L)
+    #include <latch>
+  #endif
+
 #endif
 #include <atomic>
 #include <cassert>
@@ -36,7 +44,7 @@ std::atomic<unsigned long long> failure_count {0};
 template<class ... Args>
 void log_error (char const * fmtString, Args ...args) {
   ++failure_count;
-   printf("%s", "ERROR: ");
+  printf("%s", "ERROR: ");
   printf(fmtString, args...);
   printf("%s", "\n");
   fflush(stdout);
@@ -295,6 +303,8 @@ void test_future ()
   test_future_get_value(async_member);
 }
 
+#if defined(__cplusplus) && (__cplusplus >= 202002L)
+
 void test_latch ()
 {
   std::latch latch(5);
@@ -323,9 +333,25 @@ void test_latch_release_wait ()
   log("Testing release_wait...");
   latch.arrive_and_wait(2);
 
+  log("Testing try_wait...");
+  if (!latch.try_wait())
+  {
+    log_error("try_wait did not return true even though the internal counter should have reached 0");
+  }
+
   thread1.join();
   thread2.join();
 }
+
+void test_latch_noexcept()
+{
+  static_assert(noexcept(std::latch::max()));
+  static_assert(noexcept(std::latch::latch()));
+  static_assert(noexcept(std::latch::try_wait()));
+  static_assert(noexcept(std::latch::arrive_and_wait()));
+}
+
+#endif // defined(__cplusplus) && (__cplusplus >= 202002L)
 
 #define TEST_SL_MV_CPY(ClassName) \
     static_assert(std::is_standard_layout<ClassName>::value, \
@@ -516,11 +542,14 @@ int main()
       allocated_promise.set_value(7);
     }
 
+#if defined(__cplusplus) && (__cplusplus >= 202002L)
     {
       log("Testing implementation of <latch>...");
       test_latch();
       test_latch_release_wait();
+      test_latch_noexcept();
     }
+#endif
 
     return failure_count.load() != 0;
 }

--- a/utility_scripts/Generate-StdLikeHeaders.ps1
+++ b/utility_scripts/Generate-StdLikeHeaders.ps1
@@ -62,7 +62,7 @@ param (
 $ErrorActionPreference = "Stop";
 
 # headers to be generated
-$headers = @("condition_variable", "future", "mutex", "shared_mutex", "thread")
+$headers = @("condition_variable", "future", "latch", "mutex", "shared_mutex", "thread")
 
 # ask for user input in interactive mode
 if ($Interactive) {

